File: Python/other/Maximum_Product_Subarray.py
'''
Given an array Arr that contains N integers (may be positive, negative or zero).
Find the product of the maximum product subarray.

Example 1:

Input:{6, -3, -10, 0, 2}
Output: 180
Explanation: Subarray with maximum product
is  6, -3, -10 which gives product as 180.
Example 2:

Input:{2, 3, 4, 5, -1, 0}
Output: 120
Explanation: Subarray with maximum product
is 2, 3, 4, 5 which gives product as 120.

'''
def maximum_product_subarray(arr):
    arr2 = arr[::-1]
    for i in range(1, len(arr)):
        arr[i] *= arr[i - 1] or 1
        arr2[i] *= arr2[i - 1] or 1
    return max(arr + arr2)

if __name__ == "__main__":
    arr = list(map(int,input("Enter the number: ").split()))
    print("Maximum Subarray product is ",maximum_product_subarray(arr))

'''
Time complexcity  = O(n)
space complexcity = O(1)

INPUT 
Enter the Number: 6 -3 -10 0 2

OUTPUT
Maximum Subarray product is 180
'''

File: C/other/Maximum_Product_Subarray.c
/* Find the maximum product subarray of a given array */

#include<stdio.h>
/* Function for finding minimum of two numbers */
int min(int p, int q)
{
    if (p < q)
        return p;
    else
        return q;
}

/* Function for finding maximum of two numbers */
int max(int p, int q)
{
    if (p > q)
        return p;
    else
        return q;
}

/* Function for finding maximum product subarray */
int solve()
{
    /* Input size of an array */
    int n;
    scanf("%d", &n);

    /* Declear an array*/
    int a[n];

    /* Input values in an array */
    for (int i = 0; i < n; i++)
    {
        scanf("%d" , &a[i]);
    }

    int maxend = 0, minend = 0;
    int maxupto = 0;

    /* traverse the given array */
    for (int i = 0; i < n; i++)
    {
        int temp = maxend;
        /* Update the maximum product */
        maxend = max(a[i], max(a[i] * maxend, a[i] * minend));

        /* Update the minimum product */
        minend = min(a[i], min(a[i] * temp, a[i] * minend));
        maxupto = max(maxupto, maxend);
    }

    /* Return the maximum product */
    return maxupto;
}

int main()
{
    int k = solve();

    printf("The maximum product of a subarray is %d", k);
    return 0;
}

/*
Test cases :

    Input 1 :

        8
        -4 9 -7 0 -15 6 2 -3

    Output 1 :

        The maximum product of a subarray is 540

    Input 2 :

        5
        6 -3 -10 0 2

    Output 2 :

        The maximum product of a subarray is 180

    Time complexity: O(n)
    Space Complexity: O(1)
*/
